Update Ceph RBD CSI deployment doc (#3019)
rook/rook: Documentation/ceph-csi-drivers.md

   server](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/)
3. An up and running Rook instance (see [Rook - Ceph quickstart
   guide](https://github.com/rook/rook/blob/master/Documentation/ceph-quickstart.md))

>    (e.g. kubectl create -f common.yaml, operator.yaml, cluster.yaml)

Reviewer: Point number 3 there isn't really true though is it? (I realize you didn't write it here) Getting a running rook is an effect of running steps below so a running rook isn't really a prereq.

Author: Well, actually...  my experience is that running the cluster deploy first and then running and update with the CSI components is much more reliable.  It's also what I interpreted from the old documentation even though I realize it may be possible to just use the CSI steps without deploying the cluster first.

What prompted me to update these docs was the trouble I had getting reliable, repeatable and successful deployments using the existing docs and such.

Anyway, I guess the point being if you want to remove that line that's fair enough, but I do think there's still something not quite ready to just use the csi-operator to deploy the whole thing.

Author: Cool, looks like things are kicking off just going straight to operator-with-csi.yaml  I'll let it finish up and update the doc appropriately, thanks for pointing this out!

> ## CSI Plugin Deployment and Configuration 

Reviewer 2: There is a link on line 12 that needs to be updated since this title changed

Author: Arg!  Nice catch!

kubectl apply -f cluster/examples/kubernetes/ceph/csi/rbac/cephfs/
```

> ### Create configmaps and deploy the ceph toolbox

Reviewer 3: ```suggestion
### Create configmaps and deploy the Ceph toolbox
```

*NOTE*

> The following examples assume you've cloned the [rook](https://github.com/rook/rook) repo and you're working from it's root directory. 

Reviewer 3: ```suggestion
The following examples assume you've cloned the [rook](https://github.com/rook/rook) repo and you're working from its root directory. 
```

The defaults are typically suitable for most, so we're just referencing them here

```console

> kubectl -n rook-ceph create configmap csi-rbd-config --from-file=cluster/examples/kubernetes/ceph/csi/template/rbd/

Reviewer: I thought I eliminated the need for these configmaps in my earlier PR. I can't directly see a reason they're back but maybe I'm missing something. If they are needed, could you elaborate why so I can try to eliminate them again? :-)

Author: Maybe you did, but apparently you didn't update the documentation?  Or did you and I didn't catch it? I can try running again without but like I said before; I documented the procedure that actually worked repeatably and reliably for me.

Reviewer 2: Agreed, these shouldn't be needed anymore. I knew there was a reason they looked familiar since they just went away. :)

Author: Awesome!  I'll get that removed; thanks for catching that!

kubectl -n rook-ceph create configmap csi-cephfs-config --from-file=cluster/examples/kubernetes/ceph/csi/template/cephfs/

> kubectl create -f manifests/toolbox.yaml

Author: Done

# Test RBD CSI Driver

## Create a CephBlockPool for our storage class to provision PVs from

```bash
cat << EOF | kubectl create -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: rbd
  namespace: rook-ceph
spec:
  failureDomain: host
  replicated:
    size: 3
EOF
```

## Create a Secrets object (plugin will access our Ceph cluster via this secrets file)

We need credential for two users on the Ceph Cluster; *admin* and *kubernetes*.  We can obtain these from the Operator

> POD using the ceph toolbox. 

Reviewer 3: ```suggestion
pod using the ceph toolbox. 
```

Reviewer 3: The sentence is a bit confusing, it's either from the operator or the Ceph toolbox.

*NOTE* 
You may need to create the kubernetes account on the cluster first, you can run the following to issue a get_or_create,

> if the account already exists this will just fetch it without making any modfiications to it's settings:

Reviewer 3: ```suggestion
if the account already exists this will just fetch it without making any modifications to its settings:
```

```bash

> (pod=$(kubectl get pod  -n rook-ceph -l app=rook-ceph-operator  -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph -c /var/lib/rook/rook-ceph/rook-ceph.config auth get-or-create-key client.kubernetes mon \"allow profile rbd\" osd \"profile rbd pool=rbd\"")

Reviewer 3: ```suggestion
(pod=$(kubectl get pod -n rook-ceph -l app=rook-ceph-operator -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph -c /var/lib/rook/rook-ceph/rook-ceph.config auth get-or-create-key client.kubernetes mon \"allow profile rbd\" osd \"profile rbd pool=rbd\"")
```

```

Now that we've ensured that we have our required accounts on the Ceph Cluster, we need to obtain the base64 encoded passwords to log in to these accounts and use them to create our Secrets:

```bash

> admin_key=$(pod=$(kubectl get pod  -n rook-ceph -l app=rook-ceph-operator  -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.admin -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")

Reviewer 3: ```suggestion
admin_key=$(pod=$(kubectl get pod -n rook-ceph -l app=rook-ceph-operator -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.admin -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")
```

> kubernetes_key=$(pod=$(kubectl get pod  -n rook-ceph -l app=rook-ceph-operator  -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.kubernetes -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")

Reviewer 3: ```suggestion
kubernetes_key=$(pod=$(kubectl get pod -n rook-ceph -l app=rook-ceph-operator -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.kubernetes -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")
```

```

```bash
cat << EOF | kubectl create -f -
apiVersion: v1
kind: Secret
metadata:
  name: csi-rbd-secret
  namespace: default
data:
  # Key value corresponds to a user name defined in Ceph cluster
  admin: ${admin_key}
  # Key value corresponds to a user name defined in Ceph cluster
  kubernetes: ${kubernetes_key}
  # if monValueFromSecret is set to "monitors", uncomment the
  # following and set the mon there
  #monitors: BASE64-ENCODED-Comma-Delimited-Mons
EOF
```

## Create RBD StorageClass

> Create a ``storage-class.yaml`` file, be sure to update the ``monitors`` entry to reflect the iP of your

Reviewer 3: Add an empty line before?

Author: Done

> Create a ``storage-class.yaml`` file, be sure to update the ``monitors`` entry to reflect the iP of your

Reviewer 3: ```suggestion
Create a ``storage-class.yaml`` file, be sure to update the ``monitors`` entry to reflect the IP of your
```

# Test RBD CSI Driver

## Create a CephBlockPool for our storage class to provision PVs from

```bash

> cat << EOF | kubectl create -f -

Reviewer: Isn't this cat redundant? IOW `kubectl create -f - <<EOF ...`

Author: Yup, I'll remove it

```bash
admin_key=$(pod=$(kubectl get pod -n rook-ceph -l app=rook-ceph-operator -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.admin -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")

kubernetes_key=$(pod=$(kubectl get pod -n rook-ceph -l app=rook-ceph-operator -o jsonpath="{.items[0].metadata.name}"); kubectl exec -ti -n rook-ceph ${pod} -- bash -c "ceph auth get-key client.kubernetes -c /var/lib/rook/rook-ceph/rook-ceph.config | base64")
```

```bash
cat << EOF | kubectl create -f -
apiVersion: v1
kind: Secret
metadata:
  name: csi-rbd-secret
  namespace: default
data:
  # Key value corresponds to a user name defined in Ceph cluster
  admin: ${admin_key}
  # Key value corresponds to a user name defined in Ceph cluster
  kubernetes: ${kubernetes_key}
  # if monValueFromSecret is set to "monitors", uncomment the
  # following and set the mon there
  #monitors: BASE64-ENCODED-Comma-Delimited-Mons
EOF
```

## Create RBD StorageClass

> Create a ``storage-class.yaml`` file, be sure to update the ``monitors`` entry to reflect the IP of your

Reviewer 2: Just curious, why use double ticks? Looks like it renders the same as single ticks.

Author: My vim plugin automatically makes double ticks, and because they do in fact render the same I didn't see any reason that it would be problematic.  I guess if they are annoying I can remove them though.

Reviewer 2: that's fine with double ticks, no need to change

monitors.  Using our example output above, collect the monitor IP addresses  (``kubectl -n rook-ceph get all``)

*NOTE*
Our example below uses the ``rbd`` pool we created in our first step, if you used a different name or are adding an additional pool,
be sure to modify it in the storage class 

```bash

> cat << EOF | kubectl create -f -

Reviewer 2: In documentation I think it works better to have the contents that they save to the yaml file, then use the `kubectl create` command with the pattern that was used previously. But it's probably not worth changing since this is a short-lived doc once full integration is done with rook.

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: csi-rbd
provisioner: rbd.csi.ceph.com
parameters:
    # Comma separated list of Ceph monitors
    # if using FQDN, make sure csi plugin's dns policy is appropriate.

> 	monitors: 10.108.83.214:6789,10.104.64.44:6789,10.103.170.196:6789  

Reviewer: Indentation here looks different from the surrounding blocks. (Tab?)
